tests/components/homee/test_number.py

```
@@ -0,0 +1,74 @@
+"""Test Homee nmumbers."""
+
+from unittest.mock import MagicMock, patch
+
+from syrupy.assertion import SnapshotAssertion
+
+from homeassistant.components.number import (
+    ATTR_VALUE,
+    DOMAIN as NUMBER_DOMAIN,
+    SERVICE_SET_VALUE,
+)
+from homeassistant.const import ATTR_ENTITY_ID, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry as er
+
+from . import build_mock_node, setup_integration
+
+from tests.common import MockConfigEntry, snapshot_platform
+
+
+async def test_set_value(
+    hass: HomeAssistant,
+    mock_homee: MagicMock,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test set_value service."""
+    mock_homee.nodes = [build_mock_node("numbers.json")]
+    mock_homee.get_node_by_id.return_value = mock_homee.nodes[0]
+    await setup_integration(hass, mock_config_entry)
+
+    await hass.services.async_call(
+        NUMBER_DOMAIN,
+        SERVICE_SET_VALUE,
+        {ATTR_ENTITY_ID: "number.test_number_down_position", ATTR_VALUE: 90},
+        blocking=True,
+    )
+
+    mock_homee.set_value.assert_called_once_with(1, 2, 90)
+
+
+async def test_set_value_not_editable(
+    hass: HomeAssistant,
+    mock_homee: MagicMock,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test set_value if attribute is not editable."""
+    mock_homee.nodes = [build_mock_node("numbers.json")]
+    mock_homee.get_node_by_id.return_value = mock_homee.nodes[0]
+    await setup_integration(hass, mock_config_entry)
+
+    await hass.services.async_call(
+        NUMBER_DOMAIN,
+        SERVICE_SET_VALUE,
+        {ATTR_ENTITY_ID: "number.test_number_motion_alarm_delay", ATTR_VALUE: 10000},
+        blocking=True,
+    )
+    assert not mock_homee.set_value.called
```

If the number isn't settable then a HomeAssistantError should be thrown. This will then be propagated to the UI or automation.

If the number truly isn't settable, then is should be modeled as a Sensor.

---

I was also thinking about this.
The usecase here are numbers, that might be settable or not. Example here is a Thermostat valve, which is only changeable if the Thermostat is switched to manual mode.
I decided to make the entity unavailable in that case to make it clear to the user, that it can't be changed. I had a `ServiceValdiationError` in place, but it didn't get thrown, because HA knows the entity is unavailable and the call of the service get's not executed.
So I could also skip the service call and just assert, that the entity is not available. But I decided I wanted to see that the service call definitely does not lead to an unwanted exception.

---

Yeah, got it. I think we should still be throwing an error rather than eating the error as it seems cleaner.

---

This is a little more complicated. The error isn't just "eaten". Since the entity is unavailable, HA isn't calling the function at all, but the Service-call is ignored.
A user wouldn't even have the chance to call it, since it would not be available. What I'm not sure about is, if the service could be called by an automation, that is why I kept the test the way it is.

---

> What I'm not sure about is, if the service could be called by an automation, that is why I kept the test the way it is

In your test; you have a service call that is occurring when the entity is unavailable. HA is not raising an error. The suggestion is to raise a ServiceValidationError with the appropriate error message.

---

I tried to raise an error in `set_native_value`, but it didn't raise, since the function does not get called when the entity is unavailable. So witin my integration, I don't have a cance to raise the exception.
The conclusion is, that I don't test the service on the unavailable entity, since the service does not get called anyway by HA.

---

If an entity is unavailable, HA won't call the function iirc. But I would expect a test that tests that to still raise, not by integration code, but by HA code

---

I checked it again - it doesn't throw anything - and now?

---

I just checked this out myself. HA eats the error and creates a log entry when the entity is unavailable or does not exist,. So you are correct and my feedback was incorrect.

```
Logger: homeassistant.helpers.service
Source: helpers/service.py:304
First occurred: 9:41:02 AM (3 occurrences)
Last logged: 9:41:43 AM

Referenced entities media_player.garage are missing or not currently available
```